[High Priority - 1] Advanced file mocker

satviksr opened the pull request:
The original design of the File Mapper mocker was to have the file mappings hard coded. However, due to the complexity of the File Mapper, it was not possible to accurately mock or test with this design.

This PR creates a new File Mapper which accurately mocks all the functionality of the File Mapper according to the API specification. The API Spec can be found here: https://github.com/googleinterns/filesystem-in-appsscript/blob/master/docs/FileMapper.md

This mocker is implemented as we need a way to test the FileSystem APIs before the FileMapper is ready. This is to be replaced with complete implementation. This mocker implements all the functionality as per the design of the FileMapper. The primary difference is that there is no config file, instead there is a hardcoded base directory. All file paths have to contain this base directory as prefix.

 The mocker does NOT do the following things
 1. Check MimeType, Permissions
 1. Check for duplicate files
 1. Throw errors as per design (because we are not doing checks)
 1. Prompt user in case file is not found

This is a quick implementation of the complete file mapper. It will work for all positive scenarios. I plan to write tests according to this file mapper. Tests for negative scenarios can be added once the file mapper is complete. 

Implements 
- [X] getFileId
- [X] getFolderId
- [X] hasFile
- [X] hasFolder
- [X] deleteFolder
- [X] deleteFile
- [X] createFile
- [X] createFolder
- [X] moveFile
- [X] moveFolder
- [X] copyFile
- [x] copyFolder
- [X] findFilesByPattern
- [X] findFoldersByPattern

Tests 
- [X] getFileId Tests
- [X] getFolderId Tests
- [X] hasFile Tests
- [X] hasFolder Tests
- [X] deleteFolder Tests
- [X] deleteFile Tests
- [X] createFile Tests
- [X] createFolder Tests
- [X] moveFile Tests
- [X] moveFolder Tests
- [X] copyFile Tests
- [x] copyFolder Tests
- [X] findFilesByPattern Tests
- [X] findFoldersByPattern Tests

Test Report 
![image](https://user-images.githubusercontent.com/66052573/84820303-07df8380-b037-11ea-9b31-0c8b44538f4d.png)

satviksr committed:
Complete File Mocker

satviksr commented:
@muskangarg98 can you do a quick check if everything is as expected

paneendra-google commented on FileSystem/FileMapperMocker.js:
```diff
@@ -28,10 +43,472 @@ var VBAFileMapperMocker = {
+ * 3) Throw errors as per design (because we are not doing checks)
+ * 4) Prompt user in case file is not found
+ */
+var VBAFileMapperMocker2 = {
```
Suggest dropping 2 from the name. If we already have VbaFileMapperMocker and is not going to be used then we can remove that code or make that class name as DeprecatedVbaFileMapperMocker

satviksr replied:
Fixed: Code is removed

satviksr commented on FileSystem/FileMapperMocker.js:
```diff
@@ -10,28 +25,490 @@ var USE_FILEMAPPER_MOCKER = true;
 var VBAFileMapperMocker = {
```
The old fileMapper is not removed yet to avoid conflict issues, will be removed eventually

satviksr replied:
Update: code is removed

paneendra-google commented on FileSystem/FileMapperMocker.js:
```diff
@@ -10,28 +25,490 @@ var USE_FILEMAPPER_MOCKER = true;
+  if (!fileIterator.hasNext()) {
+    throw new Error('File not found');
+  }
+  // Return file id
```
Can remove this comment as it is obvious from the code

satviksr replied:
Fixed

paneendra-google commented on FileSystem/FileMapperMocker.js:
```diff
@@ -10,28 +25,490 @@ var USE_FILEMAPPER_MOCKER = true;
+ * @param {string} localPath Local filepath of the file
+ * @return {boolean} true if the local file exists
+ */
+function hasFile(localPath) {
```
This method is almost same as getFileId, may be you can make a helper that returns  the id or null and use that in both the methods, (in the above method you can check for null and throw error), similarly for the folder ones

satviksr replied:
Fixed: Refactored to using getFileUtil()

satviksr replied:
Fixed

paneendra-google commented on FileSystem/FileMapperMocker.js:
```diff
@@ -10,28 +25,490 @@ var USE_FILEMAPPER_MOCKER = true;
+ * Deletes local file if it exists
+ * @param {string} localPath Local filepath of the file
+ */
+function deleteFile2(localPath) {
```
should not have 2 in method names (also for a few others below). If this is different from deleteFile in some way we should make that clear in the name

satviksr replied:
Fixed

paneendra-google commented on FileSystem/FileMapperMocker.js:
```diff
@@ -10,28 +25,490 @@ var USE_FILEMAPPER_MOCKER = true;
+  try {
+    var id = this.getFileId(localPath);
+    DriveApp.getFileById(id).setTrashed(true);
+  } catch (e) {
```
Should we be catching the error? we should prompt / inform the user if something went wrong?

satviksr replied:
Fixed

paneendra-google commented on FileSystem/FileMapperMocker.js:
```diff
@@ -10,28 +25,490 @@ var USE_FILEMAPPER_MOCKER = true;
+ * @param {string} targetFolderPath Destination folder path
+ */
+function copyFolder(sourceFilePaths, targetFolderPath) {
+  var targetFolderId;
```
The initial part of getting targetFolder is common to this and the function above so can make a helper function

satviksr replied:
Refactored to `getOrCreateFolderUtil`

paneendra-google commented on FileSystem/FileMapperMocker.js:
```diff
@@ -10,28 +25,490 @@ var USE_FILEMAPPER_MOCKER = true;
+  }
+  // Generate regex matching pattern
+  var pattern = pathSplit[pathSplit.length - 1];
+  var patternRegex = generateRegexFromPattern(pattern);
```
All code in the method till here are same as the one for the function above, extract out a function

satviksr replied:
Fixed: Refactored to using findByPatternUtil()

monikadhok commented on FileSystem/FileMapperMocker.js:
```diff
@@ -10,28 +25,490 @@ var USE_FILEMAPPER_MOCKER = true;
+  try {
+    var id = this.getFolderId(localPath);
+    DriveApp.getFolderById(id).setTrashed(true);
+  } catch (e) {
```
Same as above, catch the error. Notify the user what went wrong.

satviksr replied:
Fixed

monikadhok commented on FileSystem/FileMapperMocker.js:
```diff
@@ -10,28 +25,490 @@ var USE_FILEMAPPER_MOCKER = true;
+ * @param {Array} sourceFilePaths List of local folders to be moved
+ * @param {string} targetFolderPath Destination folder path
+ */
+function moveFolder(sourceFilePaths, targetFolderPath) {
```
Code for moveFile/moveFolder looks similar. Create a helper method for similar operations?

satviksr replied:
Fixed: Made helper function cloneFileUtil for moveFile and copyFile. Similarly, made cloneFolderUtil for copyFolder and moveFolder

monikadhok commented on FileSystem/FileMapperMocker.js:
```diff
@@ -10,28 +25,490 @@ var USE_FILEMAPPER_MOCKER = true;
+  var pathSplit = this.getRelativePathSplit(localPath);
+  // Get base directory
+  var root = this.getBaseDriveFolder();
+  // Get parent directory of file
```
395-400 code is used repeatedly, create a helper function.

satviksr replied:
Fixed: Refactored to using getFinalFolderUtil()

monikadhok reviewed (changes requested):
Needs code fixes.

monikadhok commented on FileSystem/FileMapperMocker.js:
```diff
@@ -1,37 +1,483 @@
+ * Mocks the FileMapper module. The mocker is implemented as we need a way to
+ * test the FileSystem APIs before the FileMapper is ready. This is to be
+ * replaced with complete implementation. This mocker implements all the
+ * functionality as per the design of the FileMapper. The primary difference is
```
functionalities?

satviksr replied:
Fixed

monikadhok commented on FileSystem/FileMapperMocker.js:
```diff
@@ -1,37 +1,483 @@
+ * @param {string} localPath Local filepath of the file
+ * @return {string} File id of the local file if available, otherwise null
+ */
+function getFileUtil(localPath) {
```
Can we move these helper methods to another class?

satviksr replied:
Fixed, moved to `FileMapperMockerUtil`

monikadhok commented on FileSystem/FileMapperMocker.js:
```diff
@@ -1,37 +1,483 @@
+ * @param {Array} pathSplit Platform agnostic path split array
+ * @return {Folder} Final folder, Null if folder not found
+ */
+function getFinalFolderUtil(pathSplit) {
```
Can we please fix these method names? getFileUtil which actually returns fileId.  Similar with getFinalFolderUtil

satviksr replied:
Fixed, refactored `getFileUtil` to `getFileIdUtil`

monikadhok commented on FileSystem/FileMapperMocker.js:
```diff
@@ -1,37 +1,483 @@
+  // Get directory
+  var root = this.getFinalFolderUtil(pathSplit);
+  if (root == null) {
+    throw new Error(localPath + ' not found');
```
Can we throw error from the helper method itself so that you don't have to do error handling in callers.

satviksr replied:
Fixed

satviksr committed:
Create FileMapperMockerUtil

paneendra-google commented on FileSystem/FileMapperMockerUtil.js:
```diff
@@ -0,0 +1,231 @@
+ * @param {string} localPath Local filepath of the folder
+ * @return {string} Folder id of the local folder if available, otherwise null
+ */
+function getFolderUtil(localPath) {
```
Generally Util is used for class names, can we use getFolderId instead?

satviksr replied:
Fixed

paneendra-google commented on FileSystem/FileMapperMockerUtil.js:
```diff
@@ -0,0 +1,231 @@
+ * @param {Array} pathSplit Platform agnostic path split array
+ * @return {Folder} Final folder
+ */
+function getFinalFolderUtil(pathSplit) {
```
Util -> Id

also in other places

satviksr replied:
Fixed

satviksr committed:
FileMapper: Remove Util Suffix

satviksr commented:
@paneendra-google, I dropped Util suffix for some of the helper functions and I renamed Util->Helper for 2 of them. 

I have raised an issues #43 so that I can fix the other modules in separate PRs.

satviksr commented:
Merging PR as PR is approved by @monikadhok and @paneendra-google

monikadhok commented on FileSystem/FileMapperMocker.js:
```diff
@@ -5,7 +5,7 @@
  * You may obtain a copy of the License at
  *
- *      https://www.apache.org/licenses/LICENSE-2.0
+ *      https://www.apache.orag/licenses/LICENSE-2.0
```
orag?

satviksr replied:
Somehow missed this, I have raised a PR to fix this typo! #54
